engine/sim/cooldown.cpp

```diff
@@ -775,3 +775,43 @@ void sc_format_to( const cooldown_t& cooldown, fmt::format_context::iterator out
 {
   fmt::format_to( out, "Cooldown {}", cooldown.name_str );
 }
+
+target_specific_cooldown_t::target_specific_cooldown_t( player_t& p, cooldown_t& base_cd ) :
+  player( &p ),
+  base_cooldown( &base_cd ),
+  name_str( base_cooldown->name() )
+{
+}
+
+cooldown_t* target_specific_cooldown_t::get_cooldown( player_t* target )
+{
+  assert( target && "Must specify a valid target to get a target_specific_cooldown" );
+
+  auto target_index = target->actor_index;
+  auto spawn_index = target->actor_spawn_index;
+
+  if ( target_cooldowns.size() <= target_index )
+    target_cooldowns.resize( target_index + 1 );
+
+  target_cooldown_t& tcd = target_cooldowns[ target_index ];
+
+  if ( !tcd.cooldown )
+  {
+    tcd.cooldown = player->get_cooldown( base_cooldown->name() + "_target_" + util::to_string( target_index ) );
+    tcd.cooldown->duration = base_cooldown->duration;
+    tcd.spawn_index = spawn_index;
+  }
+
+  if ( tcd.spawn_index != spawn_index )
+  {
+    tcd.cooldown->reset( false );
+    tcd.spawn_index = spawn_index;
+  }
+
+  return tcd.cooldown;
+}
+
+void target_specific_cooldown_t::reset()
+{
+  target_cooldowns.clear();
+}
```

Comment:
I don't think clearing the vector on reset is necessary. If I remember correctly, this would happen at start of each iteration, meaning that for simulations with many iterations this vector would have to be rebuilt every time. This probably doesn't slow things down too much, but the main purpose of keeping pointers to the cooldowns in a vector like this is to allow fast access by minimizing the number of times `player_t::get_cooldown` is called.

Reply:
I was kinda torn on this myself. The reason I plugged it in in the end was that it did seem like it would avoid a bunch of useless calls to cooldown->reset as I assumed the spawn_index values would be initially mismatched from the previous iteration. However I think you're probably right that resetting the vector would likely auto_dispose the cooldown_t* objects themselves on the player (since they aren't referenced anywhere else) and we'd have to re-create them all.

Not against removing this, just kinda seemed "technically correct" but it probably isn't strictly needed.

Reply:
That is true that there are some wasted reset calls happening, since all of the cooldowns in the player's cooldown list already run cooldown_t::reset_init when the player is reset.

I suspect that rebuilding the vector during each iteration will be more computationally expensive than those reset calls, but I'm not entirely sure, because there is a decent amount of stuff happening in cooldown_t::reset. I guess another option would be to reset the spawn indices to some negative value and to not reset the cooldowns if the old index was equal to that value.